### ISSUE TITLE
How to extract the feature embeddings

### ISSUE THREAD

**Issue by vmthanh** [dituu/Semi-Siamese-Training#2]
Hi. 
Assume that I have trained a model and could get the model weight. How could I use the model to extract the feature embeddings (512 dims) in Pytorch? Sorry Im not familiar with Pytorch framework much

```python
path = "/path/to/model/train.pt"
probe_net = MobileFaceNet(512)
probe_net = torch.nn.DataParallel(probe_net).cuda()
probe_net.load_state_dict(torch.load(path))
probe_net.eval()

## load image
img_rgb = cv2.cvtColor(cv2.imread(img_path), cv2.COLOR_BGR2RGB)
img_rgb = (img_rgb.transpose((2, 0, 1)) - 127.5) * 0.0078125
img_torch = torch.from_numpy(img_rgb.astype(np.float32))
## make evaluation
emb =  probe_net(img_torch.unsqueeze(0)).detach().cpu().numpy()
```

Although I check the shape of emb is truly 512 dims. But when I compare the cosine distance, it does not seem right. I dig in the code and see that the MobileFaceNet you use have addition layer is BN also. Im not sure this is the issue or not. 

Please help address. 

Best regards

**Comment by dituu** [comment]
> Hi.
> Assume that I have trained a model and could get the model weight. How could I use the model to extract the feature embeddings (512 dims) in Pytorch? Sorry Im not familiar with Pytorch framework much
> 
> ```python
> path = "/path/to/model/train.pt"
> probe_net = MobileFaceNet(512)
> probe_net = torch.nn.DataParallel(probe_net).cuda()
> probe_net.load_state_dict(torch.load(path))
> probe_net.eval()
> 
> ## load image
> img_rgb = cv2.cvtColor(cv2.imread(img_path), cv2.COLOR_BGR2RGB)
> img_rgb = (img_rgb.transpose((2, 0, 1)) - 127.5) * 0.0078125
> img_torch = torch.from_numpy(img_rgb.astype(np.float32))
> ## make evaluation
> emb =  probe_net(img_torch.unsqueeze(0)).detach().cpu().numpy()
> ```
> 
> Although I check the shape of emb is truly 512 dims. But when I compare the cosine distance, it does not seem right. I dig in the code and see that the MobileFaceNet you use have addition layer is BN also. Im not sure this is the issue or not.
> 
> Please help address.
> 
> Best regards

**Comment by dituu** [comment]
Hi,
 I think the dimension of emb should be squeezed in this code, that is:
`emb =  probe_net(img_torch.unsqueeze(0)).detach().cpu().numpy().squeeze(0)`
Then, you can compare the cosine distance bewtten two embeddings by: 
`np.dot(emb1, emb2.T)`

**Comment by dituu** [comment]
Hi,
I think your code is right. I am not clear why the cosine distance you said seems not right. Can you provide more detail about how to cacluate the cosine distance , or its result?

**Comment by vmthanh** [comment]
Hi.
What I mean is where in the architecture you will cut to get the face embedding. From what I saw that, for ex the MobileFaceNet, in the last layers you have conv_6_flatten, linear, bn and F.normalize
<img width="752" alt="Screenshot 2020-07-23 at 16 20 31" src="https://user-images.githubusercontent.com/8940457/88265855-cbbbd480-cd00-11ea-9806-072f4d617371.png">
If i use
```python
emb =  probe_net(img_torch.unsqueeze(0)).detach().cpu().numpy()
```
From what I understand is that it will calculate from the beginning to the F.normalize and get the input. Should this the correct way? 
But if cutting a conv_6_flatten only, the embedding will be different 
```python
import torch.nn as nn
def get_feature_of_mobile_face_net(net, layer_name, x):
    x = x.unsqueeze(0)
    for name, module in net._modules.items():
        x = module(x)
        if name == layer_name:
            
            return x
```
```python
emb = get_feature_of_mobile_face_net(probe_net,"conv_6_flatten", face_semi)
```
So what is the correct way? In addition, some other details that I would like to know or it is much appreciated if you could provide more training code and validation like:
- I tried with AttentionNet from you also. And again, in which layer of Attention we should use to get the embedding vectors?
- What is the color order of image (BGR or RGB) ? 
- Is it required to normalise like what I wrote in the above comment? 
- How to train in our own dataset? What I do currently is that I replicate lmdb_utils.py (since I dont know how to use lmdb and where to the training file) and write my own python file. What it does here is at one step, it will randomly read a pair of two image and push it to train model.

Regarding to your question, I had my own dataset containing some pair of images, then I use trained model to do the embedding and calculate the cosine similarity: 
```python
def cal_cosine_similarity(emb1, emb2):
    face_cos_mat = sklearn.metrics.pairwise.cosine_similarity([emb1], [emb2])
    face_cosine = float(face_cos_mat[0][0])
    return face_cosine
```

Again, Im very thankful with your work and your contribution. 

Best.

**Comment by dituu** [comment]
Hi, 
we get the embedding vectors through the complete backbone network.  We don't cut the architecture. So, the first way is correct. In evaluation, we extract features from a batch data as follows:
```
with torch.no_grad(): 
    for batch_idx, image in enumerate(data_loader):
        image = image.to(device)
        feature = model(image).cpu().numpy()
```
It's just like we extract features from the gallery netwotk during the training. 

- As for AttentionNet, it follows the same way. 

- Besides, the image preprocessing should be consistent in training and testing. In our implementation, we set the color order of images as RGB, and normalize the images the same as your wrote. 

- In each training iteration, a batch of ids is sampled (the ids should be non-overlapped in prototype queue), and two images are randomly sampled from each id. On shallow data, each id only has two images, so we directly input them to training model.

Thanks for following our work, we are working to provide a more general and complete implementation.

**Comment by dituu** [comment]
> Hi.
> What I mean is where in the architecture you will cut to get the face embedding. From what I saw that, for ex the MobileFaceNet, in the last layers you have conv_6_flatten, linear, bn and F.normalize
> <img alt="Screenshot 2020-07-23 at 16 20 31" width="752" src="https://user-images.githubusercontent.com/8940457/88265855-cbbbd480-cd00-11ea-9806-072f4d617371.png">
> If i use
> 
> ```python
> emb =  probe_net(img_torch.unsqueeze(0)).detach().cpu().numpy()
> ```
> 
> From what I understand is that it will calculate from the beginning to the F.normalize and get the input. Should this the correct way?
> But if cutting a conv_6_flatten only, the embedding will be different
> 
> ```python
> import torch.nn as nn
> def get_feature_of_mobile_face_net(net, layer_name, x):
>     x = x.unsqueeze(0)
>     for name, module in net._modules.items():
>         x = module(x)
>         if name == layer_name:
>             
>             return x
> ```
> 
> ```python
> emb = get_feature_of_mobile_face_net(probe_net,"conv_6_flatten", face_semi)
> ```
> 
> So what is the correct way? In addition, some other details that I would like to know or it is much appreciated if you could provide more training code and validation like:
> 
> * I tried with AttentionNet from you also. And again, in which layer of Attention we should use to get the embedding vectors?
> * What is the color order of image (BGR or RGB) ?
> * Is it required to normalise like what I wrote in the above comment?
> * How to train in our own dataset? What I do currently is that I replicate lmdb_utils.py (since I dont know how to use lmdb and where to the training file) and write my own python file. What it does here is at one step, it will randomly read a pair of two image and push it to train model.
> 
> Regarding to your question, I had my own dataset containing some pair of images, then I use trained model to do the embedding and calculate the cosine similarity:
> 
> ```python
> def cal_cosine_similarity(emb1, emb2):
>     face_cos_mat = sklearn.metrics.pairwise.cosine_similarity([emb1], [emb2])
>     face_cosine = float(face_cos_mat[0][0])
>     return face_cosine
> ```
> 
> Again, Im very thankful with your work and your contribution.
> 
> Best.

**Comment by vmthanh** [comment]
Hi.
Noted. Thanks for your reply. 
Im looking forward to seeing your updates in the future. 
Best regards.
